interp/run.go

```diff
@@ -1125,55 +1125,69 @@ func methodByName(value reflect.Value, name string) reflect.Value {
 func call(n *node) {
 	goroutine := n.anc.kind == goStmt
 	var method bool
-	value := genValue(n.child[0])
+	c0 := n.child[0]
+	value := genValue(c0)
 	var values []func(*frame) reflect.Value
 
 	recvIndexLater := false
 	switch {
-	case n.child[0].recv != nil:
+	case c0.recv != nil:
 		// Compute method receiver value.
-		values = append(values, genValueRecv(n.child[0]))
+		values = append(values, genValueRecv(c0))
 		method = true
-	case len(n.child[0].child) > 0 && n.child[0].child[0].typ != nil && isInterfaceSrc(n.child[0].child[0].typ):
+	case len(c0.child) > 0 && c0.child[0].typ != nil && isInterfaceSrc(c0.child[0].typ):
 		recvIndexLater = true
-		values = append(values, genValueBinRecv(n.child[0], &receiver{node: n.child[0].child[0]}))
+		values = append(values, genValueBinRecv(c0, &receiver{node: c0.child[0]}))
 		value = genValueBinMethodOnInterface(n, value)
 		method = true
-	case n.child[0].action == aMethod:
+	case c0.action == aMethod:
 		// Add a place holder for interface method receiver.
 		values = append(values, nil)
 		method = true
 	}
 
-	numRet := len(n.child[0].typ.ret)
+	numRet := len(c0.typ.ret)
 	variadic := variadicPos(n)
 	child := n.child[1:]
 	tnext := getExec(n.tnext)
 	fnext := getExec(n.fnext)
 
 	// Compute input argument value functions.
 	for i, c := range child {
+		var arg *itype
+		if variadic >= 0 && i >= variadic {
+			arg = c0.typ.arg[variadic].val
+		} else {
+			arg = c0.typ.arg[i]
+		}
 		switch {
 		case isBinCall(c, c.scope):
 			// Handle nested function calls: pass returned values as arguments.
 			numOut := c.child[0].typ.rtype.NumOut()
 			for j := 0; j < numOut; j++ {
 				ind := c.findex + j
-				values = append(values, func(f *frame) reflect.Value { return f.data[ind] })
+				if !isInterfaceSrc(arg) || isEmptyInterface(arg) {
+					values = append(values, func(f *frame) reflect.Value { return f.data[ind] })
+					continue
+				}
+				values = append(values, func(f *frame) reflect.Value {
+					return reflect.ValueOf(valueInterface{value: f.data[ind]})
+				})
```

Comment:
I think we should add a "break" statement after this line, because I've just realized that it is not possible to nest a function that returns more than one argument. Or maybe we should catch this earlier if NumOut is > 1.
For example, consider this (invalid) program:

```
package main

import (
	"errors"
)

func tworets() (string, error) {
	return "hello", errors.New("an error")
}

func takesfour(i int, s string, err error, ss string) {
	println(i)
	println(s)
	println(err)
	println(ss)
}

func main() {
	takesfour(7, tworets(), "world")
}
```

```
% go run ./sample.go 
# command-line-arguments
./sample.go:19:11: not enough arguments in call to takesfour
./sample.go:19:22: multiple-value tworets() in single-value context
% yaegi run ./sample.go 
run: ./sample.go:19:15: cannot use func() (string,error) as type string
```

So fortunately it looks like we also get an error in yaegi in that case, but I think adding a break here would make the intention clearer and would be safer anyway. WDYT?

or am I missing something?

Reply:
I had forgotten about this legitimate case:

```
package main

func tworets() (string, int) {
	return "hello", 4
}

func takestwo(s string, i int) {
	println(s)
	println(i)
}

func main() {
	takestwo(tworets())
}
```

which is allowed because the number of arguments is exactly the same as the number of returned values.
But I've just tried and it still works with the breaks I proposed. And I think it is more efficient as it does less work.

Reply:
nvm, I had misunderstood.